src/tcache.c

```diff
@@ -416,44 +482,61 @@ tcache_init(tsd_t *tsd, tcache_t *tcache, void *avail_stack) {
 
 	ticker_init(&tcache->gc_ticker, TCACHE_GC_INCR);
 
-	size_t stack_offset = 0;
 	assert((TCACHE_NSLOTS_SMALL_MAX & 1U) == 0);
 	memset(tcache->bins_small, 0, sizeof(cache_bin_t) * SC_NBINS);
 	memset(tcache->bins_large, 0, sizeof(cache_bin_t) * (nhbins - SC_NBINS));
+
 	unsigned i = 0;
+	bool overflow_fixup = false;
+	uintptr_t stack_cur = (uintptr_t)avail_stack;
 	for (; i < SC_NBINS; i++) {
 		tcache->lg_fill_div[i] = 1;
-		stack_offset += tcache_bin_info[i].ncached_max * sizeof(void *);
-		/*
-		 * avail points past the available space.  Allocations will
-		 * access the slots toward higher addresses (for the benefit of
-		 * prefetch).
-		 */
-		tcache_small_bin_get(tcache, i)->avail =
-		    (void **)((uintptr_t)avail_stack + (uintptr_t)stack_offset);
+		cache_bin_t *bin = tcache_small_bin_get(tcache, i);
+		if (tcache_bin_init(bin, i, &stack_cur)) {
+			assert(!overflow_fixup);
+			overflow_fixup = true;
+			/* Reinit the current bin after padding. */
+			i--;
+		}
 	}
 	for (; i < nhbins; i++) {
-		stack_offset += tcache_bin_info[i].ncached_max * sizeof(void *);
-		tcache_large_bin_get(tcache, i)->avail =
-		    (void **)((uintptr_t)avail_stack + (uintptr_t)stack_offset);
+		cache_bin_t *bin = tcache_large_bin_get(tcache, i);
+		if (tcache_bin_init(bin, i, &stack_cur)) {
+			assert(!overflow_fixup);
+			overflow_fixup = true;
+			/* Reinit the current bin after padding. */
+			i--;
+		}
 	}
-	assert(stack_offset == stack_nelms * sizeof(void *));
+
+	/* Sanity check that the whole stack is used. */
+	size_t stack_offset = stack_cur - (uintptr_t)avail_stack;
+	assert(stack_offset + total_stack_padding == total_stack_bytes +
+	    overflow_fixup ? cache_bin_max_offset : 0);
 }
 
 /* Initialize auto tcache (embedded in TSD). */
 bool
 tsd_tcache_data_init(tsd_t *tsd) {
 	tcache_t *tcache = tsd_tcachep_get_unsafe(tsd);
-	assert(tcache_small_bin_get(tcache, 0)->avail == NULL);
-	size_t size = stack_nelms * sizeof(void *);
+	assert(tcache_small_bin_get(tcache, 0)->cur_ptr.ptr == NULL);
 	/* Avoid false cacheline sharing. */
-	size = sz_sa2u(size, CACHELINE);
-
+	size_t size = sz_sa2u(total_stack_bytes, CACHELINE);
 	void *avail_array = ipallocztm(tsd_tsdn(tsd), size, CACHELINE, true,
 	    NULL, true, arena_get(TSDN_NULL, 0, true));
 	if (avail_array == NULL) {
 		return true;
 	}
+	if (tcache_bin_lowbits_overflowable(avail_array, total_stack_bytes)) {
+		idalloctm(tsd_tsdn(tsd), avail_array, NULL, NULL, true, true);
+		/* Add padding so the overflowable range can be skipped. */
+		size = sz_sa2u(size + cache_bin_max_offset, CACHELINE);
+		avail_array = ipallocztm(tsd_tsdn(tsd), size, CACHELINE, true,
+		    NULL, true, arena_get(TSDN_NULL, 0, true));
+		if (avail_array == NULL) {
+			return true;
+		}
+	}
```

Comment:
Add `assert(!tcache_bin_lowbits_overflowable(...))`?

Reply:
Here the total space + total size may still appear overflowable; but in that case there will be enough padding for us to skip a bin's space that may overflow the low bits. The assertion here https://github.com/jemalloc/jemalloc/pull/1592/files#diff-777e740de5d78972a2742086c8dc7f67R514 should verify that case.